pkg/controller/gc_controller.go

```diff
@@ -103,104 +176,74 @@ func (c *gcController) Run(ctx context.Context, workers int) error {
-			deletionFailure = true
+			errs = append(errs, errors.Wrapf(err, "error deleting snapshot %s", volumeBackup.SnapshotID))
 		}
 	}
 
-	// If applicable, delete everything in the backup dir in object storage *before* deleting the API object
-	// because otherwise the backup sync controller could re-sync the backup from object storage.
-	if deleteBackupFiles {
-		logContext.Info("Removing backup from object storage")
-		if err := c.backupService.DeleteBackupDir(c.bucket, backup.Name); err != nil {
-			logContext.WithError(err).Error("Error deleting backup")
-			deletionFailure = true
-		}
+	log.Info("Removing backup from object storage")
+	if err := c.backupService.DeleteBackupDir(c.bucket, backup.Name); err != nil {
+		errs = append(errs, errors.Wrap(err, "error deleting backup from object storage"))
 	}
 
-	logContext.Info("Getting restore API objects referencing backup")
 	if restores, err := c.restoreLister.Restores(backup.Namespace).List(labels.Everything()); err != nil {
-		logContext.WithError(errors.WithStack(err)).Error("Error getting Restore API objects")
+		log.WithError(errors.WithStack(err)).Error("Error listing restore API objects")
 	} else {
 		for _, restore := range restores {
-			if restore.Spec.BackupName == backup.Name {
-				logContext.WithField("restore", kube.NamespaceAndName(restore)).Info("Removing Restore API object referencing Backup")
-				if err := c.restoreClient.Restores(restore.Namespace).Delete(restore.Name, &metav1.DeleteOptions{}); err != nil {
-					logContext.WithError(errors.WithStack(err)).WithField("restore", kube.NamespaceAndName(restore)).
-						Error("Error deleting Restore API object")
-				}
+			if restore.Spec.BackupName != backup.Name {
+				continue
 			}
-		}
-	}
 
-	if deletionFailure {
-		logContext.Warning("Backup will not be deleted due to errors deleting related object storage files(s) and/or volume snapshots")
-		return
-	}
+			restoreLog := log.WithField("restore", kube.NamespaceAndName(restore))
 
-	logContext.Info("Removing Backup API object")
-	if err := c.backupClient.Backups(backup.Namespace).Delete(backup.Name, &metav1.DeleteOptions{}); err != nil {
-		logContext.WithError(errors.WithStack(err)).Error("Error deleting Backup API object")
-	}
-}
-
-// garbageCollectBackups checks backups for expiration and triggers garbage-collection for the expired
-// ones.
-func (c *gcController) garbageCollectBackups(backups []*api.Backup, expiration time.Time, deleteBackupFiles bool) {
-	for _, backup := range backups {
-		if backup.Status.Expiration.Time.After(expiration) {
-			c.logger.WithField("backup", kube.NamespaceAndName(backup)).Info("Backup has not expired yet, skipping")
-			continue
+			restoreLog.Info("Deleting restore referencing backup")
+			if err := c.restoreClient.Restores(restore.Namespace).Delete(restore.Name, &metav1.DeleteOptions{}); err != nil {
+				restoreLog.WithError(errors.WithStack(err)).Error("Error deleting restore")
```

Comment:
Do you think this should just be an error so we consider the GC of the backup a failure?

Reply:
eh, in my opinion no -- I'm not too worried about orphaned restores (since they don't cost any cloud resources), plus that would have the result of leaving the backup API object around, despite its tarball/etc. potentially having been deleted, which could make a user believe it's still restorable.

Reply:
wfm